File: lib/textHelpers.ts
/**
 * Text helper utilities for string manipulation
 */

/**
 * Maximum allowed string length to prevent DoS attacks (10MB of characters)
 */
const MAX_STRING_LENGTH = 10 * 1024 * 1024;

/**
 * Regular expression to detect control characters that could be used for injection
 */
const CONTROL_CHARS_REGEX = /[\x00-\x08\x0B-\x0C\x0E-\x1F\x7F]/;

/**
 * Capitalizes the first letter of each word in a string while preserving original spacing.
 * 
 * This function performs the following:
 * - Splits the input string by whitespace
 * - Capitalizes the first letter of each word
 * - Preserves original spacing and punctuation
 * - Handles Unicode characters properly
 * - Includes security validations to prevent injection attacks
 * 
 * @param input - The string to capitalize. Can be null, undefined, or empty.
 * @returns A new string with the first letter of each word capitalized
 * 
 * @throws {TypeError} If input is not a string (when not null/undefined)
 * @throws {RangeError} If input exceeds maximum allowed length
 * @throws {Error} If input contains control characters
 * 
 * @example
 * ```typescript
 * capitalizeWords("hello world") // Returns: "Hello World"
 * capitalizeWords("hello  world") // Returns: "Hello  World" (preserves spacing)
 * capitalizeWords("café naïve") // Returns: "Café Naïve"
 * capitalizeWords("don't stop") // Returns: "Don't Stop"
 * capitalizeWords("") // Returns: ""
 * capitalizeWords(null) // Returns: ""
 * ```
 * 
 * @security
 * - Validates input type to prevent injection attacks
 * - Rejects strings with control characters
 * - Enforces maximum string length to prevent DoS
 * - Does not use eval() or dynamic code execution
 * - No logging of input to prevent exposure of sensitive data
 */
export function capitalizeWords(input: string | null | undefined): string {
  // Handle null and undefined inputs gracefully
  if (input === null || input === undefined) {
    return '';
  }

  // Validate input type
  if (typeof input !== 'string') {
    throw new TypeError('Input must be a string');
  }

  // Handle empty string
  if (input.length === 0) {
    return '';
  }

  // Security: Check for maximum length to prevent DoS
  if (input.length > MAX_STRING_LENGTH) {
    throw new RangeError(`Input string exceeds maximum allowed length of ${MAX_STRING_LENGTH} characters`);
  }

  // Security: Reject strings with control characters
  if (CONTROL_CHARS_REGEX.test(input)) {
    throw new Error('Input contains invalid control characters');
  }

  // Split by spaces while preserving the original delimiters
  // This regex captures both words and the spaces between them
  const parts = input.split(/(\s+)/);

  // Process each part
  const result = parts.map(part => {
    // If this is a whitespace segment, preserve it as-is
    if (/^\s+$/.test(part)) {
      return part;
    }

    // If this is an empty part, return as-is
    if (part.length === 0) {
      return part;
    }

    // Capitalize the first character and append the rest
    // Use toLocaleUpperCase for proper Unicode handling
    const firstChar = part.charAt(0).toLocaleUpperCase();
    const restOfWord = part.slice(1);

    return firstChar + restOfWord;
  });

  return result.join('');
}

/**
 * Capitalizes the first letter of each word in a string with custom delimiter support.
 * 
 * This is an advanced version that allows specifying custom delimiters for word boundaries.
 * 
 * @param input - The string to capitalize
 * @param delimiters - Array of delimiter characters (default: whitespace only)
 * @returns A new string with the first letter of each word capitalized
 * 
 * @throws {TypeError} If input is not a string (when not null/undefined)
 * @throws {RangeError} If input exceeds maximum allowed length
 * @throws {Error} If input contains control characters
 * 
 * @example
 * ```typescript
 * capitalizeWordsWithDelimiters("hello-world", ["-"]) // Returns: "Hello-World"
 * capitalizeWordsWithDelimiters("hello_world_test", ["_"]) // Returns: "Hello_World_Test"
 * ```
 */
export function capitalizeWordsWithDelimiters(
  input: string | null | undefined,
  delimiters: string[] = []
): string {
  // Handle null and undefined inputs gracefully
  if (input === null || input === undefined) {
    return '';
  }

  // Validate input type
  if (typeof input !== 'string') {
    throw new TypeError('Input must be a string');
  }

  // Handle empty string
  if (input.length === 0) {
    return '';
  }

  // Security: Check for maximum length to prevent DoS
  if (input.length > MAX_STRING_LENGTH) {
    throw new RangeError(`Input string exceeds maximum allowed length of ${MAX_STRING_LENGTH} characters`);
  }

  // Security: Reject strings with control characters
  if (CONTROL_CHARS_REGEX.test(input)) {
    throw new Error('Input contains invalid control characters');
  }

  // If no custom delimiters, fall back to standard implementation
  if (delimiters.length === 0) {
    return capitalizeWords(input);
  }

  // Escape special regex characters in delimiters
  const escapedDelimiters = delimiters.map(d => d.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'));
  
  // Create regex pattern that captures delimiters and words
  const pattern = new RegExp(`([${escapedDelimiters.join('')}\\s]+)`, 'g');
  
  // Split by delimiters while preserving them
  const parts = input.split(pattern);

  // Process each part
  const result = parts.map(part => {
    // If this is a delimiter segment or whitespace, preserve it as-is
    if (pattern.test(part) || part.length === 0) {
      return part;
    }

    // Capitalize the first character and append the rest
    const firstChar = part.charAt(0).toLocaleUpperCase();
    const restOfWord = part.slice(1);

    return firstChar + restOfWord;
  });

  return result.join('');
}

File: lib/text_helpers.py
"""
Text manipulation helper functions.

This module provides utility functions for text processing with security
considerations and proper input validation.
"""

import re
from typing import Optional


# Maximum string length to prevent DoS attacks (10MB)
MAX_STRING_LENGTH = 10 * 1024 * 1024


class StringValidationError(ValueError):
    """Raised when string validation fails."""
    pass


def capitalize_words(text: Optional[str]) -> str:
    """
    Capitalize the first letter of each word in a string.
    
    This function capitalizes the first character of each word while preserving
    the original spacing and structure of the input string. It handles Unicode
    characters properly and includes security safeguards against malicious input.
    
    Args:
        text: The input string to process. Can be None (returns empty string).
    
    Returns:
        A new string with the first letter of each word capitalized.
        Returns empty string for None or empty input.
    
    Raises:
        StringValidationError: If input exceeds maximum length or contains
                              invalid null bytes.
        TypeError: If input is not a string or None.
    
    Security Considerations:
        - Input length is limited to prevent DoS attacks
        - Null byte injection is prevented
        - No dynamic code execution is used
        - Control characters are preserved but not interpreted
    
    Examples:
        >>> capitalize_words("hello world")
        'Hello World'
        >>> capitalize_words("hello  world")
        'Hello  World'
        >>> capitalize_words("don't stop")
        "Don't Stop"
        >>> capitalize_words("café naïve")
        'Café Naïve'
        >>> capitalize_words("")
        ''
        >>> capitalize_words(None)
        ''
    """
    # Handle None input
    if text is None:
        return ""
    
    # Type validation
    if not isinstance(text, str):
        raise TypeError(f"Expected string or None, got {type(text).__name__}")
    
    # Handle empty string
    if not text:
        return ""
    
    # Security: Check maximum length to prevent DoS
    if len(text) > MAX_STRING_LENGTH:
        raise StringValidationError(
            f"String length exceeds maximum allowed length of {MAX_STRING_LENGTH} bytes"
        )
    
    # Security: Prevent null byte injection
    if '\x00' in text:
        raise StringValidationError("String contains null bytes")
    
    # Process the string character by character to preserve all spacing
    result = []
    # Track if we're at the start of a new word
    capitalize_next = True
    
    for char in text:
        if char.isspace():
            # Preserve all whitespace characters
            result.append(char)
            # Next non-space character should be capitalized
            capitalize_next = True
        elif char.isalpha():
            # Capitalize if we're at the start of a word
            if capitalize_next:
                result.append(char.upper())
                capitalize_next = False
            else:
                result.append(char.lower())
        else:
            # Handle punctuation, numbers, and other characters
            # These don't reset the word boundary for cases like "don't"
            result.append(char)
            # Only reset capitalize_next for certain punctuation
            # that typically ends a word (not apostrophes)
            if char in '.,;:!?-—–':
                capitalize_next = True
    
    return ''.join(result)


def sanitize_control_characters(text: str, replace_with: str = ' ') -> str:
    """
    Remove or replace control characters from text.
    
    This is a helper function for cases where control characters need to be
    sanitized before processing.
    
    Args:
        text: The input string to sanitize.
        replace_with: Character to replace control characters with (default: space).
    
    Returns:
        Sanitized string with control characters replaced.
    
    Examples:
        >>> sanitize_control_characters("hello\x07world")
        'hello world'
    """
    if not isinstance(text, str):
        raise TypeError(f"Expected string, got {type(text).__name__}")
    
    # Replace control characters (except newline, tab, carriage return)
    # with the replacement character
    return re.sub(r'[\x00-\x08\x0B-\x0C\x0E-\x1F\x7F]', replace_with, text)